habitat-lab/habitat/core/batch_renderer.py

```diff
@@ -0,0 +1,344 @@
+                    sensor_spec.uuid
+                ][env_index]
+            # Post-process sim sensor output using lab sensor interface.
+            # The same lab sensors are re-used for all environments.
+            processed_obs: Observations = self._sensor_suite.get_observations(
+                env_observations
+            )
+            for key, value in processed_obs.items():
+                env_observations[key] = value
+            output.append(env_observations)
+        return output
+
+    def draw_observations(
+        self, sensor_spec: BackendSensorSpec
+    ) -> Union[np.ndarray, "Tensor"]:
+        r"""
+        Draw observations for all environments.
+
+        :param sensor_spec: Habitat-sim sensor specifications.
+        :return: A numpy ndarray in GPU-to-CPU mode, or a torch tensor in GPU-to-GPU mode.
+        """
+        draw_fn: Callable = (
+            self.draw_observations_gpu_to_gpu
+            if self._gpu_gpu
+            else self.draw_observations_gpu_to_cpu
+        )
+        return draw_fn(sensor_spec)
+
+    def draw_observations_gpu_to_cpu(
+        self, sensor_spec: BackendSensorSpec
+    ) -> np.ndarray:
+        r"""
+        Draw observations for all environments.
+        Copies sensors output from GPU memory into CPU ndarrays, during which the thread is blocked.
+
+        :param sensor_spec: Habitat-sim sensor specifications.
+        :return: ndarray containing renders.
+        """
+        # TODO: Currently only one color sensor is supported.
+        if sensor_spec.sensor_type == habitat_sim.SensorType.COLOR:
+            if self._gpu_to_cpu_images is None:
+                # Allocate the transfer buffers
+                self._gpu_to_cpu_images = []
+                storage = mn.PixelStorage()
+                storage.alignment = 2
+                self._gpu_to_cpu_buffer = np.empty(
+                    (
+                        self._num_envs,
+                        sensor_spec.resolution[0],
+                        sensor_spec.resolution[1],
+                        sensor_spec.channels,
+                    ),
+                    dtype=np.uint8,
+                )
+
+                for env_idx in range(self._num_envs):
+                    # Create image view for writing into buffer from Magnum
+                    env_img_view = mn.MutableImageView2D(
+                        mn.PixelFormat.RGBA8_UNORM,
+                        [sensor_spec.resolution[1], sensor_spec.resolution[0]],
+                        self._gpu_to_cpu_buffer[env_idx],
+                    )
+                    self._gpu_to_cpu_images.append(env_img_view)
+
+                # Flip the transfer buffer view vertically for presentation
+                self._gpu_to_cpu_buffer = np.flip(
+                    self._gpu_to_cpu_buffer.view(), axis=1
+                )
+        else:
+            raise NotImplementedError
+
+        # Render
+        self._replay_renderer.render(self._gpu_to_cpu_images)
```

Comment:
Is there an assert somewhere to catch if we're trying to use this with more than one color sensor per env?